### ISSUE TITLE
ld.lld: error: undefined symbol: __aeabi_read_tp

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#1485]
looks like in linux-next, a new link failure are occurring for 32b arm.

thumb2: https://github.com/ClangBuiltLinux/continuous-integration2/runs/3964216542?check_suite_focus=true

```
Error: ld.lld: error: undefined symbol: __aeabi_read_tp
```
@compnerd mentioned some familiarity with this symbol and that this might be a bug in LLVM.  @arndb reported this in yesterday's meeting.

**Comment by nathanchance** [comment]
I think there was some discussion in our meeting yesterday that `__builtin_thread_pointer()` was not working properly for Thumb2?

**Comment by nickdesaulniers** [comment]
@ardbiesheuvel mentioned [this](https://git.kernel.org/pub/scm/linux/kernel/git/ardb/linux.git/commit/?id=50596b7559bf226bb35ad55855ee979453ec06a1) which introduces `__builtin_thread_pointer`.

**Comment by arndb** [comment]
Here is a trivial reproducer https://godbolt.org/z/57f9vfvrE

**Comment by nickdesaulniers** [comment]
@ardbiesheuvel points out that `-mtp=cp15 -marm --target=arm-linux-gnueabi- -O2` works, `-mtp=cp15 -mthumb --target=arm-linux-gnueabi- -O2` does not.

**Comment by nickdesaulniers** [comment]
Adding `-mtp=cp15` adds to the IR the following function attribute: `"target-features"=+read-tp-hard,+...`.

**Comment by ardbiesheuvel** [comment]
https://reviews.llvm.org/D112600

**Comment by nathanchance** [comment]
I opened an LLVM PR so that we can try to get this into LLVM 13.0.1: https://llvm.org/pr52329

**Comment by nickdesaulniers** [comment]
Fixed in llvm-14: https://reviews.llvm.org/rGd7e089f2d6a5cd5f283a90ab29241d20d4fc3ed1.

We could ask about getting it back ported to clang 13.0.1, though we'll probably need patches on the kernel side to avoid this feature on clang-12 and older?

**Comment by nathanchance** [comment]
Yes, I opened a PR above so that we can get this merged into 13.0.1 but also yes, there should be a kernel side fix so that the build does not break.

**Comment by nickdesaulniers** [comment]
@ardbiesheuvel 's kernel patch: https://lore.kernel.org/linux-arm-kernel/20211028083527.1057158-1-ardb@kernel.org/

**Comment by nathanchance** [comment]
The kernel patch has been merged into mainline: https://git.kernel.org/linus/c1e42efacb9bb4ae873e9b1cf249fa4fb6ef7f84

The LLVM patch should be picked up from LLVM 13.0.1 based on Peter's response on the LLVM bug: https://llvm.org/pr52329

I will close this once the LLVM patch lands in the `release/13.x` branch.

**Comment by nathanchance** [comment]
`release/13.x` cherry-pick: https://github.com/llvm/llvm-project/commit/ed38280d006c91d7098b0a02e160835ee09fd08a
